### ISSUE TITLE
[css-fonts-4] The spec should not allow UAs to ignore user-installed fonts

### ISSUE THREAD

**Issue by r12a** [w3c/csswg-drafts#5421]
10.3. Preinstalled Fonts and User-Installed Fonts
https://drafts.csswg.org/css-fonts-4/#preinstalled-and-user-installed-fonts

> User Agents may choose to ignore User-Installed Fonts for the purpose of the Font Matching Algorithm.

!!!! This will create consternation for a great many international users around the world. A huge number of people use languages which are not supported, or are not supported well, by preinstalled fonts. The almost universal approach, currently, for people who use less common or endangered languages on the Web is to find a page where someone has created a font (and often keyboard) that you are expected to download in order to see or create content.  (Those fonts are often used by applications other than browsers, too.)

Furthermore, quite often there are also particular font styles that content authors want to see used, often contrastively, and there's a likelihood that a user will have installed additional fonts (such as using a Kano font for Nigerian ajami text, or a Mool font style to distinguish Khmer headings, etc.) in order to make the text readable/locally relevant. For many languages, the system may provide a Noto font, but usually only the one (and generally the Noto fonts don't reflect the design users would want to see for their text, since it focuses on producing Noto-harmonised, large, sans-serif, monoline glyphs).

Of course, this plays into the issues related to privacy, but if the motivation for the sentence is that, it should say so, at least. But that discussion is not yet concluded, so in the meantime the i18n WG would like to see that sentence removed.

**Comment by xfq** [comment]
If I remember correctly, the font based fingerprinting discussions started last year, but this text has been in the spec for at least [2 years](https://github.com/w3c/csswg-drafts/commit/a6ef2d9420509ec3a9ac9c9f899d4d0588381381#). Perhaps it's for another reason?

**Comment by astearns** [comment]
@litherum knows better, but my understanding is that this sentence was added mostly to describe reality - in that Safari had already implemented this behavior and continues to behave this way. The subsequent fingerprinting discussion has happened in part to figure out how to address problems with this behavior.

Two of the suggestions we have considered are:

1) Allow some User-Installed Fonts to match for particular languages where many web pages depend on them
2) User agents that have this font-matching behavior MUST provide UI to let the user opt-in to User-Installed Font matching

Would either or both of these requirements make the behavior more acceptable to the i18n WG?

**Comment by r12a** [comment]
> Safari had already implemented this behavior

But Safari is only one of the 3 main browser engines, so i don't believe the spec should reflect that approach yet, because it doesn't represent reality.  (And btw, i am no longer able to use or recommend Safari for a deal of multilingual content for this reason, esp. when it concerns support for endangered scripts or scripts that people are promoting for wider use. Which is a nuisance, since it writes off use on iPads and iPhones as well as desktop.) 

(1) above sounds problematic because i don't believe it's possible to create and maintain in a timely fashion a centralised list that will respond to the needs of the content developers, especially for endangered and long-tail scripts. Installing fonts currently avoids that problem, of course.

(2) Is a given, to my mind, if any restrictions are to be placed on ability to render user-installed fonts, but it adds complications for non-computer-literate users that it must be easy for them to understand how to work around.

However, both of those are solutions to a problem that we don't yet have, since the majority of major browser engines do not restrict visibility of user-installed fonts.  So if we are to describe reality, we should remove that sentence.  If we are to describe some future possibility, which to my mind is not yet certain enough to include in the spec, then such a statement must be accompanied by another that says that users must be able to work around the restriction when needed.

**Comment by astearns** [comment]
One argument that (2) could be an acceptable workaround is that someone who can figure out how to install a font can likely figure out how to change a browser preference.

And one of the options we have discussed for (1) is to flip the maintenance requirement. So we would allow a UA to ignore user-installed fonts ONLY for languages where it can be demonstrated that available OS fonts are adequate for displaying web content.

**Comment by litherum** [comment]
Web fonts are the only sure way for a web author to use these less-supported languages. Naming a specific font and hoping the user has it installed is less likely to work than using web fonts.

There’s a trade-off here. International support is clearly important, but privacy is also definitely important. Different UA’s policies have different implications for different users. I chose the word “may” here on purpose, simply to describe the trade-off. There are many different points on the spectrum between blocking fonts for privacy and allowing fonts for internationalization, and all of those points have reasonable arguments for them, based on the specific considerations of the UA and their users.

The sentence “User Agents may choose to ignore User-Installed Fonts” is *true* (de facto). We shouldn’t remove it from the spec.

**Comment by svgeesus** [comment]
> Web fonts are the only sure way for a web author to use these less-supported languages. 

In general, using a Web font is more reliable *provided the license allows such use*. This is not the case for all fonts, including some which are "free to use" but are required to be distributed in an intact zip file with a readme and license, for example.

> Naming a specific font and hoping the user has it installed is less likely to work than using web fonts.

It is a very poor approach for Latin, agreed.

It is the _standard_ approach in some linguistic communities, because they came onto the Web well before Web fonts were a realistic thing to use. Thus, user-agents which suddenly stop working on content which has worked for one or two decades will simply be seen as broken.

**Comment by r12a** [comment]
> we would allow a UA to ignore user-installed fonts ONLY for languages where it can be demonstrated that available OS fonts are adequate for displaying web content.

What is the criterion against which you'd measure adequacy?  I'm not sure how that would work.

Here are some additional thoughts about things that concern me...

> Web fonts are the only sure way for a web author to use these less-supported languages.

I'm curious about how you reached that conclusion. Do you base this on experience, or is it a logical deduction based on the assumption in the sentence "Naming a specific font and hoping the user has it installed is less likely to work than using web fonts."?  I ask because, as someone who has worked with a wide range of less-supported languages on a daily basis for several years, it doesn't match the reality that i experience. In fact, it appears to be the opposite.

Less-supported languages typically have only a small number of freely available Unicode fonts, and those are often obtained by users from the Web.  Users expect to use them with applications such as text editors  or for texting, as well as for Web pages, and they are therefore often packaged along with keyboards.  Webfonts are not useful for those scenarios. It's very rare for such packages or download locations to serve the user with a webfont.  Webfonts are a tool for content authors, not for users.

It may be possible to persuade content authors to create and use webfonts for new content, but they aren't going to help with the content that's currently out there.

Not that webfonts are a panacea, anyway.  They require additional bandwidth, when people using less-supported languages may find that a problem.  And for documents containing multiple languages that can sometimes mount up.


Access to user-installed fonts can also give the user the flexibility to set preferred fonts as the default.  For example, i've just been writing about the Javanese script as used for the Javanese language. There is a Noto font for Javanese, although it has some bugs still, but it also uses a noto-harmonised style that doesn't look like normal Javanese book fonts (although a recent redesign now at least makes it possible to actually read the letters).  Otherwise, there is very little to choose from wrt Javanese fonts.  That I know of, the only font that actually works well and looks the way i'd expect is Yogyakarta.  It's all rights reserved, so i can't as a content author create a webfont for it. I can, however, prepend it to my CSS font-family values so that those who also happen to have it can also obtain the benefits of a well designed font, rather that having to settle for an alternative system font.

At the very least, browsers shouldn't block content authors from trying out different fonts they have installed while developing their content.  Even if they are going to send the content out with a single webfont, they need to be able to experiment quickly with alternatives in order to choose that font.  Compiling system fonts into webfonts and incorporating into the content for testing out the effect is way too slow.  Localhost and file URLs should be exempt from these restrictions, since there is no privacy risk there anyway.

**Comment by astearns** [comment]
@litherum the only way I personally would consider your user stylesheet hack to be sufficient is if the user agent automated it for the user. It would be fine to have it as the *mechanism* for opting in, but there would need to be relatively-simple UI presented to the user.

**Comment by litherum** [comment]
I'm particularly troubled by @astearns's comment in https://github.com/w3c/csswg-drafts/issues/5421#issuecomment-709425330 about the original edit being premature.

Previously, the spec said nothing about the set of exposed fonts. Adding the sentence "User Agents may choose to ignore User-Installed Fonts" didn't actually change anything -user agents were _already_ free to ignore these fonts because nothing was defined before. The edit takes something that was entirely undefined and simply calls out the fact that it was undefined. I wouldn't call this "premature."

One valid criticism would be that the original edit doesn't describe the fact that there are many different places in the middle of the spectrum of [forbid all user-installed fonts, allow all user-installed fonts] that UAs can choose, and that these places are (generally) _more_ valuable than either of the extremes. Saying "UAs may pick one extreme" doesn't capture this nuance. Fixing this is an improvement to the spec, and is the resolution the group just agreed on.

**Comment by litherum** [comment]
I made a pull request with proposed text. Please take a look! https://github.com/w3c/csswg-drafts/pull/5625

**Comment by astearns** [comment]
The spec said nothing, but there was (and is) a common understanding that user agents could and would match user-installed fonts. Yes, this was under-specified. But we had rough interoperability on user-installed fonts until WebKit changed. Adding the possibility that user-installed fonts could be ignored was a *new* thing.

Since the edit was pointed out, we have had lots of discussions about the benefits and drawbacks of what’s in the spec. I think there is consensus that allowing a user agent to ignore user-installed fonts with no usable opt-in is too extreme. I do think the current MAY statement was premature, because it explicitly allows that extreme behavior.

So I support the edit you have made. And I support continuing to discuss usable opt-in scenarios so we can get back to having a normative statement we can all agree on.

**Comment by litherum** [comment]
@astearns 

> there is consensus that allowing a user agent to ignore user-installed fonts with no usable opt-in is too extreme.

I don't agree with this. WebKit is certainly interested in refining its balance between internationalization and privacy, but that isn't the same as saying no UA should ever pick either of the extremes. Some UAs have very strong internationalization goals, and other browsers have very strong privacy goals.

I would agree with you if some content was becoming unreadable _for no benefit_. But there is a benefit here, and so it becomes a cost/benefit analysis problem. All options should be on the table, even if we expect that most UAs won't choose one of the options.

**Comment by r12a** [comment]
For me, i think it comes down to this: User agents may pick a position along the internationalisation<->privacy tension line to propose to the user by default, but they shouldn't take decisions for the user that the user can't _easily_ change.

And in the case of fonts, although an all-on/all-off switch might be useful too, i think this generally means the user being able to choose which user-installed fonts work on a font-by-font basis.

One problem is that the user is not likely to be aware what font the content author is trying to apply.  That's why i had suggested elsewhere an approach like we have for cookies, that warns the user if the page is trying to use a font that is not already enabled by them, and asks then remembers whether they want to enable it for this and future pages.

**Comment by svgeesus** [comment]
@r12a there was a proposal by @felipeerias did you see it?

https://github.com/w3c/csswg-drafts/issues/4497#issuecomment-763459971

It seems to satisfy the "user can easily change" aspect that you mentioned; I particularly liked the per-site/everywhere option. It also makes fingerprinting slower, even if the user has authorized a large number of fonts.

**Comment by svgeesus** [comment]
So, do the changes made in https://github.com/w3c/csswg-drafts/commit/b0ac8dd0ee00450ead86efdb79e30d0781ea6f28 satisfy the competing constraints on privacy vs. digital access, or is there more to do?

**Comment by svgeesus** [comment]
The specifi wording here being 

> In addition, User Agents may have additional facilities for fine-tuning this balance, 
such as interfaces to allow users to explicitly make 
certain fonts available or unavailable. 

Re-reading this, it misses out "to particular domains or to all of them" suggested by @felipeerias and I think that should be added.

**Comment by svgeesus** [comment]
However this part **is** normative: [10.3. Preinstalled Fonts and User-Installed Fonts](https://drafts.csswg.org/css-fonts-4/#preinstalled-and-user-installed-fonts):

> User agents may choose to ignore User-Installed Fonts for the purpose of the [Font Matching Algorithm](https://drafts.csswg.org/css-fonts-4/#font-matching-algorithm).

My suggestion therefore is to _strike that sentence_, on the grounds that we already say the set of installed fonts is undefined and that we already give non-normative guidance on privacy-vs-Internationalization trade-offs.

**Comment by litherum** [comment]
If I understand correctly, striking that sentence doesn’t affect UA behavior, conformance, or test pass rate. Therefore, it would be an editorial change.

Given that it’s an editorial change, I think it’s valuable to remind the user in different points in the spec document about things they may be interested in. Therefore, I think it would be better to reword that sentence to better encapsulate the privacy-fidelity tradeoff, rather than strike it entirely.

**Comment by svgeesus** [comment]
> Therefore, I think it would be better to reword that sentence to better encapsulate the privacy-fidelity tradeoff, rather than strike it entirely.

Do you have suggested wording, which @r12a could then review? I feel we could be close to closing this long-running issue.

**Comment by litherum** [comment]
I think what makes sense is:

1. It doesn't make any sense to discuss this in the taxonomy section. I think the taxonomy section should just simply describe, well, taxonomy, rather than its implications on other algorithms. There fore, I think we should replace both "The set of [installed fonts](https://drafts.csswg.org/css-fonts-4/#installed-fonts) available in the [Font Matching Algorithm](https://drafts.csswg.org/css-fonts-4/#font-matching-algorithm) is explicitly undefined" and "User agents may choose to ignore User-Installed Fonts for the purpose of the [Font Matching Algorithm](https://drafts.csswg.org/css-fonts-4/#font-matching-algorithm)" with "Please see [Font Matching Algorithm](https://drafts.csswg.org/css-fonts-4/#font-matching-algorithm) for a description of how user-installed fonts may interact with the font matching algorithm.

2. Add a paragraph to 5. Font Matching Algorithm that describes the tradeoff. Something like:

> The set of [installed fonts](https://drafts.csswg.org/css-fonts-4/#installed-fonts) available in the [Font Matching Algorithm](https://drafts.csswg.org/css-fonts-4/#font-matching-algorithm) is explicitly undefined.
> 
> Note: UAs may choose to expose all installed fonts to the web, regardless of how that font was installed; doing so is likely to have good internationalization properties for users whose primary language is not supported by fonts shipping with their operating system. UAs may alternatively choose to not expose user-installed fonts to aid privacy on the web; for the set of installed fonts a user has installed is often used as a tracking vector to track users across the web. UAs may choose a hybrid approach, where some user-installed fonts are exposed for internationalization, but others aren't. UAs are expected to make informed decisions on which fonts they expose to the web, so as to balance between internationalization and privacy.

**Comment by svgeesus** [comment]
I agree about moving much of this from taxonomy, but prefer to use [the actual text](https://github.com/w3c/csswg-drafts/commit/b0ac8dd0ee00450ead86efdb79e30d0781ea6f28) (currently in taxonomy) which was proposed and agreed; your proposal misses out several points that were important in the discussion and requested by several people.

**Comment by svgeesus** [comment]
@r12a @xfq could you check the changes [made here](https://github.com/w3c/csswg-drafts/commit/5ead211946b37c5db19e4dd55d1c30510299ee49) satisfy the concern raised in this issue?

**Comment by r12a** [comment]
The large initial change block seems to be heading in a useful direction.  However, the 3rd option (hybrid approach) assumes that the UA implementer is able to choose adequate fonts to be exposed for a language on behalf of the user.

> UAs are expected to make informed decisions on which fonts they expose to the web, so as to balance between internationalization and privacy.

I don't see how they can possibly do that in a way that meets user's requirements.  I still think that if a UA is going to prevent use of installed fonts, then the user needs to be able to unblock those fonts if they wish.  The user also needs to be able to view text in fonts that the UA implementer has never even heard of.  And it also doesn't address the need i mentioned earlier for content developers to be able to view content in whatever font they want in situations that are not security risks, such as reading a file from their own hard disk or local server while creating content.

(Slightly tangiential, but indicative of the problems we are opening the door for here is that users should also be able to overwrite or delete any font that is pre-installed, too.  For example, Kashmiri users can't currently use Safari because they need access to very recent versions of nastaliq fonts like Noto in order to correctly write their language (and fix the non-standardised and ugly hacks that users have resorted to so far), but macOS won't allow a new version to delete or overwrite the pre-installed version.  This is really preventing users from creating readable content. [See also an example](https://typo.social/@webi18n@w3c.social/111489082833818738) of the problems caused for _many_ writing systems for users of Firefox on the mac, for similar reasons.)

**Comment by svgeesus** [comment]
Thanks for the detailed feedback @r12a, let's keep working to resolve this.

**Comment by svgeesus** [comment]
@r12a 
>  I still think that if a UA is going to prevent use of installed fonts, then the user needs to be able to unblock those fonts if they wish.

Please take a look at https://github.com/w3c/csswg-drafts/commit/d7193b93677c12324b3ebba9b202d4f5a1f89c94 where this is made explicit.

> Slightly tangiential, but indicative of the problems we are opening the door for here is that users should also be able to overwrite or delete any font that is pre-installed, too. 

Not that tangential; the wording I just added talks about removing as well as adding fonts from the initial default set provided by the UA. Removing from a set used for rendering seemed a better way to express it than "must allow uninstalling".

**Comment by r12a** [comment]
LGTM, thanks.

**Comment by svgeesus** [comment]
@pes10k could you look over

- [5. Font Matching Algorithm](https://drafts.csswg.org/css-fonts-4/#font-matching-algorithm)
- [10.3. Preinstalled Fonts and User-Installed Fonts](https://drafts.csswg.org/css-fonts-4/#preinstalled-and-user-installed-fonts)

from a privacy perspective?

**Comment by pes10k** [comment]
This text does not address the privacy concern. Unless im looking at the wrong version, this text just makes the existing privacy problem more explicit; that the defined behavior could be used to harm user privacy and its up to vendors to figure out how to implement the spec w/o the privacy harm.

In order for the privacy concern to be addressed, it it should be the case that a correct and complete implementation of the standard ensures that the defined functionality is privacy preserving; not just that a vendor could implement it in such a way that it could be privacy preserving.

I appreciate the importance of the I18n issues here, but I think its critical to find a way to address those concerns _without_ risking user privacy.

FWIW, heres another suggestion you all may have already considered and discarded (but maybe not!): discarding the OS vs installed distinction, and instead limiting the number of font-faces that an site / eTLD+1 can reference within the lifetime a storage partition. Like i said, maybe you all have considered and discarded the idea, but this kind of "budgeting approach" might be successful for fonts in a way I dont think its practical for addressing fingerprinting concerns in general (for example, benign font-face selection might be consistent in a way that a sites overall "set of Web APIs used" is not).

**Comment by astearns** [comment]
I’ve created a [new issue](https://github.com/w3c/csswg-drafts/issues/11753) to explore making the current spec text more prescriptive
